src/algebra/non_unital_alg_hom.lean

```diff
@@ -0,0 +1,171 @@
+rfl
+
+@[simp, norm_cast] lemma coe_to_linear_map (f : non_unital_alg_hom R A B) :
+  ((f : A →ₗ[R] B) : A → B) = f :=
+rfl
+
+@[simp, norm_cast] lemma coe_to_mul_hom (f : non_unital_alg_hom R A B) :
+  ((f : mul_hom A B) : A → B) = f :=
+rfl
+
+@[norm_cast] lemma coe_linear_map_mk (f : non_unital_alg_hom R A B) (h₁ h₂ h₃) :
+  ((⟨f, h₁, h₂, h₃⟩ : non_unital_alg_hom R A B) : A →ₗ[R] B) = ⟨f, h₁, h₂⟩ :=
+by { ext, refl, }
+
+@[norm_cast] lemma coe_mul_hom_mk (f : non_unital_alg_hom R A B) (h₁ h₂ h₃) :
+  ((⟨f, h₁, h₂, h₃⟩ : non_unital_alg_hom R A B) : mul_hom A B) = ⟨f, h₃⟩ :=
+by { ext, refl, }
+
+@[simp] lemma map_smul (f : non_unital_alg_hom R A B) (c : R) (x : A) :
+  f (c • x) = c • f x :=
+f.to_linear_map.map_smul c x
+
+@[simp] lemma map_add (f : non_unital_alg_hom R A B) (x y : A) :
+  f (x + y) = (f x) + (f y) :=
+f.to_linear_map.map_add x y
+
+@[simp] lemma map_mul (f : non_unital_alg_hom R A B) (x y : A) :
+  f (x * y) = (f x) * (f y) :=
+f.to_mul_hom.map_mul x y
+
+@[simp] lemma map_zero (f : non_unital_alg_hom R A B) : f 0 = 0 :=
+f.to_linear_map.map_zero
+
+instance : has_zero (non_unital_alg_hom R A B) :=
+⟨{ map_mul' := by simp,
+   .. (0 : A →ₗ[R] B)}⟩
+
+instance : has_one (non_unital_alg_hom R A A) :=
+⟨{ map_mul' := by simp,
+   .. (1 : A →ₗ[R] A) }⟩
+
+@[simp] lemma coe_zero : ((0 : non_unital_alg_hom R A B) : A → B) = 0 := rfl
+
+@[simp] lemma coe_one : ((1 : non_unital_alg_hom R A A) : A → A) = id := rfl
+
+lemma zero_apply (a : A) : (0 : non_unital_alg_hom R A B) a = 0 := rfl
+
+lemma one_apply (a : A) : (1 : non_unital_alg_hom R A A) a = a := rfl
+
+instance : inhabited (non_unital_alg_hom R A B) := ⟨0⟩
+
+/-- The composition of morphisms is a morphism. -/
+def comp (f : non_unital_alg_hom R B C) (g : non_unital_alg_hom R A B) : non_unital_alg_hom R A C :=
+{ .. mul_hom.comp (f : mul_hom B C) (g : mul_hom A B),
+  .. linear_map.comp (f : B →ₗ[R] C) (g : A →ₗ[R] B) }
+
+@[simp, norm_cast] lemma coe_comp (f : non_unital_alg_hom R B C) (g : non_unital_alg_hom R A B) :
+  (f.comp g : A → C) = (f : B → C) ∘ (g : A → B) :=
+rfl
+
+lemma comp_apply (f : non_unital_alg_hom R B C) (g : non_unital_alg_hom R A B) (x : A) :
+  f.comp g x = f (g x) :=
+rfl
+
+/-- The inverse of a bijective morphism is a morphism. -/
+def inverse (f : non_unital_alg_hom R A B) (g : B → A)
+  (h₁ : function.left_inverse g f) (h₂ : function.right_inverse g f) :
+  non_unital_alg_hom R B A :=
+{ .. mul_hom.inverse (f : mul_hom A B) g h₁ h₂,
+  .. linear_map.inverse (f : A →ₗ[R] B) g h₁ h₂ }
+
+@[simp] lemma coe_inverse (f : non_unital_alg_hom R A B) (g : B → A)
+  (h₁ : function.left_inverse g f) (h₂ : function.right_inverse g f) :
+  (inverse f g h₁ h₂ : B → A) = g :=
+rfl
+
+end non_unital_alg_hom
+
+namespace alg_hom
+
+variables [comm_semiring R] [semiring A] [semiring B] [algebra R A] [algebra R B]
+
+instance non_unital_alg_hom.has_coe : has_coe (A →ₐ[R] B) (non_unital_alg_hom R A B) :=
+⟨λ f, { map_smul' := f.map_smul, .. f, }⟩
```

Comment:
For consistency with all the other hom conversions, I'd probably lean towards having an explicit `to_non_unital_alg_hom` definition too.